openmdao/core/driver.py

```diff
@@ -310,16 +314,66 @@ def _setup_driver(self, problem):
                     sz = sizes[owner, i]
 
                 if vname in dv_set:
-                    dv_dict[vname] = (owner, sz)
+                    remote_dv_dict[vname] = (owner, sz)
+
+                # Note that design vars are not distributed.
                 elif distributed:
                     idx = model._var_allprocs_abs2idx['nonlinear'][vname]
                     dist_sizes = model._var_sizes['nonlinear']['output'][:, idx]
-                    dist_con_dict[vname] = (idx, dist_sizes)
+
+                    # Determine which indices are on our proc.
+                    rank = model.comm.rank
+                    size = dist_sizes.size
+                    offsets = np.cumsum(dist_sizes)
+
+                    if vname in self._responses:
+                        resp_dict = self._responses[vname]
+                        indices = resp_dict['indices']
+
+                        if indices is not None:
+                            local_indices = []
+                            true_sizes = np.zeros(size, dtype=INT_DTYPE)
+                            for index in indices:
+                                if index < 0:
+                                    # Support for negative indices. Convert to positive index.
+                                    index = index + np.sum(dist_sizes)
+                                irank = np.argwhere(offsets > index)[0][0]
+                                true_sizes[irank] += 1
+                                if rank == irank:
+                                    new_index = index - offsets[irank] + dist_sizes[irank]
+                                    local_indices.append(new_index)
+
+                            indices = local_indices
+                            dist_sizes = true_sizes
+
+                        dist_con_dict[vname] = (indices, dist_sizes)
```

Comment:
Should there be a check around this to make sure this is actually an constraint and not an objective?

Reply:
Refactored so that we only have responses.